[Breaking] `RuntimeIdentifier` No Longer Implies `SelfContained` Apps by Default

nagilson opened the pull request:
# `RuntimeIdentifier` No Longer Implies `SelfContained` Apps by Default
- RIDS no longer imply Self-Contained for TFM >= 8.0
- Warns for apps that didn't set SC but set RIDs in TFM < 8 that they need to set SC explicitly to get the same behavior with TFM 8.0. Warning will go away after SC is set, either to true or false.
- PublishTrimmed, PublishSingleFile, and PublishAot will keep the old behavior of implying SC with 8+ TFMs, though they will only do so for publish which is a breaking change. PublishReadyToRun will not maintain this behavior and you will need to specify SelfContained explicitly to use it with PublishReadyToRun in TFMs above 7.

- Remove the old error that occurred when `-r` given without `-self-contained` as it did not catch property in project scenarios, and it's redundant with the breaking change warning. We also have a more understandable default now, (setting one property does not add another that is not necessary.)
- Update IL Link Tests and Publish* tests that expected the old behavior. Talked with that team, we don't want to add an implicit `SelfContained` there even if it will break customers as it limits the design potential in the future for these properties.
- Talked to Blazor WASM who would be broken by this, they have created work items to add `SelfContained` so customers aren't broken. I modified their tests as well to pass by adding `SelfContained`. 

This is the current VS behavior and we designed this in a NET 8 breaking changes list. It will help us align with VS behavior.

# Tested Behavior:
* RidNoLongerImpliesSelfContainedForTFM8+
* SelfContainedPlusRuntimeIdentifierDoesNotTriggerEightTFMFDDBreakingChangeWarning

# Questions:
- Why does `TargetFrameworkInference` sometimes not run? I accounted for that, but don't understand the contexts.

nagilson committed:
Make the breaking change of RID -> No Longer -> SC

nagilson committed:
Update XLF and remove breaking change error.

nagilson committed:
Improve the error message quality.

nagilson committed:
Fix a test which did not properly check if app was SelfContained or not

Update ILLink Tests to use SelfContained explicitly

Repair Publish* Tests that will be broken and now need to add SC

Update publish trimmed test to expect self contained to be explicitly specified

Update Publish* related tests to add SC

PublishSingleFile test adds SC now

IL Link and Publish* Tests now add SelfContained where a RuntimeIdentifier used to infer it automaticall per .NET 8 breaking change

nagilson committed:
Update global property flow test to not expect rid to infer sc

nagilson committed:
Update blazor wasm tests to add SC with the new RID change

Make blazor tests not give SC if already defined. Fix additional tests expecting old RID -> SC behavior

Blazor WASM Update test assets

Blazor WASM Tests Updated to add SelfContained when a RuntimeIdentifier is specified to keep old behavior after .NET 8 breaking change.

GlobalPropertyFlowTests now expect RID to not Infer SC

* Note that the failure was moved from 1 function to another ... technically the old test failed because it was relying on RID setting SC which would cause a failure as one EXE was SC but the other wasn't. With the new change, that no longer occured

Meanwhile, the other test relied on the properties flowing the RID. When the RID flowed, SC would be inferred, so the referenced project would be an EXE. But now it's not inferred, so this test should instead fail.

nagilson committed:
Fix test with missing framework for RID breaking change

nagilson committed:
Update Framework References tests to add SelfContained per breaking change in .net 8

nagilson committed:
Update WPF tests that expected RID to give SC

nagilson committed:
Update framework reference invalid rid and trimmable tests to no longer expect SC to give a RID.

nagilson committed:
Repair some publish profile tests to add SelfContained and update another blazor test asset to add SC

nagilson committed:
Update dotnet build tests to add SelfContained explicitly when needed for 8 breaking change

nagilson committed:
Update publish profile tests to add SC explicitly per .net 8 breaking change

nagilson committed:
Fix EXE global property flow tests to add SelfContained now when a RID is given.

nagilson committed:
Blazor: Add SelfContained in SDK tasks project instead of the test projects

nagilson committed:
Revert "Blazor: Add SelfContained in SDK tasks project instead of the test projects"

This reverts commit 552b1b35411ae921c5403a4b5602ede8d7fbd657.

nagilson committed:
Update warning and remove old warning

Remove the old self-contained warning as it would be redundant with the breaking change warning and only occurred with -r.

Co-authored-by: richlander <rlander@microsoft.com>

nagilson committed:
Update newly added breaking change test to also assert SelfContained is correct regardless of the value

sbomer reviewed (approved):
ILLink test changes LGTM. We could probably clean up the tests by adding SelfContained to the project file, but no need to do so in this change.

nagilson commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:
```diff
@@ -229,7 +230,7 @@ public void ILLink_respects_global_TrimMode(string targetFramework, string trimM
 
             var publishCommand = new PublishCommand(testAsset);
-            publishCommand.Execute($"/p:RuntimeIdentifier={rid}", "/p:PublishTrimmed=true").Should().Pass();
+            publishCommand.Execute($"/p:RuntimeIdentifier={rid}", "/p:SelfContained=true", "/p:SelfContained=true", "/p:PublishTrimmed=true").Should().Pass();
```
Whoops, sorry. I missed that mistake. Thanks for catching it.

nagilson committed:
Don't add `SelfContained` twice in the IL Link test, as this was a mistake

Co-authored-by: Sven Boemer <sbomer@gmail.com>

nagilson committed:
Don't add `SelfContained` twice in the test, as this was a mistake

Co-authored-by: Sven Boemer <sbomer@gmail.com>

nagilson committed:
Don't add `SelfContained` twice in the test, as this was a mistake

Co-authored-by: Sven Boemer <sbomer@gmail.com>

nagilson committed:
Improve test code quality by adding SelfContained in less places where possible (IL Link) and other minor code style improvements from other PR feedback

nagilson committed:
Add SelfContained where needed still in IL Link

nagilson committed:
Fix remaining IL Link Tests with the new change to not add SelfContained everywhere so we only add it in tests that should pass

nagilson committed:
merge with main

richlander commented:
Is this still blocked?

nagilson commented:
Nope, our meeting from last week cleared up any blockers on this one. It's in the pipeline to take care of soon :)

nagilson commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:
```diff
@@ -1555,15 +1558,17 @@ public void Build_respects_IsTrimmable_property(string targetFramework, bool isE
             var buildCommand = new BuildCommand(testAsset);
             // IsTrimmable enables analysis warnings during build
-            buildCommand.Execute()
+            buildCommand.Execute("/bl")
```
```suggestion
            buildCommand.Execute()
```

nagilson committed:
Imply SC for some publish properties as we decided to in the partner sync

nagilson committed:
Merge by updating string resx

nagilson committed:
update global property flow test on merge to have the breaking change expectations

nagilson committed:
Remove the binlog added to a test

nagilson committed:
update xlf and add test for publish property inference behavior

nagilson committed:
Merge branch 'nagilson-rid-no-more-sc-2.0' of https://github.com/nagilson/sdk into nagilson-rid-no-more-sc-2.0

nagilson committed:
fix bad msbuild syntax

nagilson committed:
fix path too long in test

nagilson committed:
make the tests pass by recollecting that sc still infers rid

nagilson committed:
fix global property flow tests to acc for new behaviour again

nagilson committed:
Make the framework reference flow in a test by adding self contained explicitly

nagilson commented on src/Assets/TestProjects/TrimmedAppWithReferences/App/App.csproj:
```diff
@@ -5,6 +5,7 @@
     <TargetFramework>$(CurrentTargetFramework)</TargetFramework>
     <PublishTrimmed>true</PublishTrimmed>
     <EnableTrimAnalyzer>false</EnableTrimAnalyzer>
+    <SelfContained>True</SelfContained>
```
```suggestion

```

nagilson committed:
remove selfcontained added to test lib because it should get inferred

nagilson commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:
```diff
@@ -149,7 +163,9 @@ Copyright (c) .NET Foundation. All rights reserved.
-    <SelfContained Condition="'$(SelfContained)' == '' and '$(RuntimeIdentifier)' != ''">true</SelfContained>
+    <!-- Breaking change in .NET 8: Projects with 8.0+ TFMS will no longer have RuntimeIdentifier imply SelfContained. Note that PublishReadyToRun will imply SelfContained in these versions. -->
+    <SelfContained Condition="'$(SelfContained)' == '' and '$(RuntimeIdentifier)' != '' and
+                              '$(_TargetFrameworkVersionWithoutV)' &lt; '8.0'">true</SelfContained>
```
I dont think we need to check if _TargetFrameworkVersionWithoutV is empty by this point

nagilson replied:
Eh better be safe than sorry

dsplaisted replied:
I'd suggest switching to `$([MSBuild]::VersionLessThan($(_TargetFrameworkVersionWithoutV), '8.0'))`.

nagilson replied:
Good point, I didnt know about that. Though we need to keep the empty string checks because that function will fail with an empty string

nagilson commented on src/Tests/Microsoft.NET.Build.Tests/GlobalPropertyFlowTests.cs:
```diff
@@ -248,11 +235,13 @@ public void TestGlobalPropertyFlowInSolution(bool passSelfContained, bool passRu
         }
 
-        private static void ValidateProperties(TestAsset testAsset, TestProject testProject, bool expectSelfContained, bool expectRuntimeIdentifier, string targetFramework = null, string expectedRuntimeIdentifier = "")
+        private static void ValidateProperties(TestAsset testAsset, TestProject testProject, bool expectSelfContained, bool expectRuntimeIdentifier, string thisTargetFramework = null, string expectedRuntimeIdentifier = "")
```
The parameter rename here is because the IDE was getting confused when doing type checking on which targetFramework property was which.

nagilson committed:
merge string resx with artifact  path

nagilson committed:
Merge branch 'nagilson-rid-no-more-sc-2.0' of https://github.com/nagilson/sdk into nagilson-rid-no-more-sc-2.0

nagilson committed:
add empty targetframeworkversion check

nagilson committed:
update xlf

nagilson committed:
Revert "remove selfcontained added to test lib because it should get inferred"

This reverts commit e16385b69ad96443b5efc207ac5fe47e38a9dd11.

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs:
```diff
@@ -383,6 +385,75 @@ public void It_can_publish_runtime_specific_apps_with_library_dependencies_self_
+            var buildCommand = new DotnetBuildCommand(Log);
+            var commandResult = buildCommand
+                .WithWorkingDirectory(testAsset.Path)
+                .Execute("/bl");
```
I think we should avoid producing binlogs during CI test runs, they just take up extra space (and it would be a lot of extra space if all the tests were doing it).  When I'm debugging locally I'll add the `/bl` parameter if needed, and sometimes I'll accidentally check it in.  But I think we should avoid doing it intentionally and remove it when we notice it.

nagilson replied:
Thanks for explaining. I agree.

nagilson committed:
remove binlogs from tests

nagilson committed:
use the msbuild version comparison function instead of lt and rt in runtimeidentifierinference

nagilson committed:
fix container test which had different bl syntax

baronfel reviewed (approved):
Approved the changes to the containers files on behalf of @dotnet/sdk-container-builds-maintainers - this change is actually something we're really excited about, because this old default meant that most folks were making large images by default!

nagilson committed:
add the empty tfm check back as the msbuild function will fail if given empty string

nagilson commented:
@dsplaisted @joeloff  @richlander  PTAL at reviewing when possible. Id like to try to push this into preview 4 if it looks good and I think theres really only one or two days left for that. Thank you!

baronfel committed:
Update container tests to push to correct net8 base images

With RID no longer forcing SelfContained, the base image for many tests changed.
In addition, for 8 previews we track very aggressively and we're currently ahead of
the image tags that are available - SDK-facing tests therefore must relax a bit and
pin runtimes to a slightly earlier version.

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GlobalPropertyFlowTests.cs:
```diff
@@ -86,7 +86,7 @@ public void TestGlobalPropertyFlowToLibrary(bool passSelfContained, bool passRun
 
             bool buildingSelfContained = passSelfContained || passRuntimeIdentifier;
```
This should probably renamed or the logic reworked, since passing a RuntimeIdentifier doesn't imply that the project will be self-contained.  Same with other tests here.

nagilson replied:
I changed it to `appIsSelfContainedOrRuntimeSpecific` which is longer and could be better, but I think it does make it clearer

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GlobalPropertyFlowTests.cs:
```diff
@@ -182,18 +165,22 @@ public void TestGlobalPropertyFlowToMultitargetedProject(bool passSelfContained,
+            string identifier = passSelfContained.ToString() + "_" + passRuntimeIdentifier;
 
-            bool buildingSelfContained = passSelfContained || passRuntimeIdentifier;
+                var testAsset = Build(passSelfContained, passRuntimeIdentifier, identifier: identifier);
```
Looks like the indentation might be off here.

dsplaisted commented on src/Tests/Microsoft.NET.Publish.Tests/RuntimeIdentifiersTests.cs:
```diff
@@ -271,6 +271,58 @@ public void PublishRuntimeIdentifierOverridesUseCurrentRuntime()
+        [InlineData("PublishTrimmed", false)]
+        public void SomePublishPropertiesInferSelfContained(string property, bool useFrameworkDependentDefaultTargetFramework)
+        {
+            // Note: there is a bug with PublishAot I think where this test will fail for Aot if the testname is too long. Do not make it longer.
```
In that case you can set the TestProject's Name to something shorter.

dsplaisted reviewed (approved):
Generally looks good.  I didn't look at all the test changes for things like global property flow and referencing Exes in as much detail as would be nice.

It sounds like we want to fix part of the breaking change for blazor-wasm, I think we should probably do that in this PR rather than working around it in the tests.

All the test changes highlight that this is a pretty high-impact change.  It helps that it's restricted to projects targeting .NET 8 or higher, but we probably need to make sure to get the word out about this and expect some impact from the change. @richlander @baronfel

nagilson committed:
Improve clarity of warning message for breaking change

Co-authored-by: Daniel Plaisted <dsplaisted@gmail.com>

nagilson commented on src/Tests/Microsoft.NET.Build.Containers.IntegrationTests/CreateNewImageTests.cs:
```diff
@@ -138,13 +138,13 @@ public void Tasks_EndToEnd_With_EnvironmentVariable_Validation()
         File.WriteAllText(Path.Combine(newProjectDir.FullName, "Program.cs"), $"Console.Write(Environment.GetEnvironmentVariable(\"GoodEnvVar\"));");
 
-        new DotnetCommand(_testOutput, "build", "--configuration", "release", "/p:runtimeidentifier=linux-x64")
+        new DotnetCommand(_testOutput, "build", "--configuration", "release", "/p:runtimeidentifier=linux-x64", $"/p:RuntimeFrameworkVersion=8.0.0-preview.3.23174.8")
```
I agree, I believe @baronfel did this as a temporary patch to get the container tests to pass as there is no structure in place to make those tests pass while the preview 4 and preview 5 container bits havent been built.

baronfel replied:
Echoing this - the container tests are in kind of an odd space (at least for the preview timeframe). The SDK version being used to _build_ containers carries along a certain RuntimeFrameworkVersion implicitly, but the containers made available on `mcr.microsoft.com` lag behind in terms of the runtimes installed on them (by design - they are only made in the run-ups to each preview release) and so in the SDK we tend to make framework-dependent apps that we actually _can't_ run in containers.  We (containers team) need a better strategy for this and so I added these properties to 'pin' the runtimes for projects targeting net8.0 as a temporary measure.

baronfel replied:
I went ahead and extracted this comment^ to https://github.com/dotnet/sdk/issues/32099 for the containers team to follow-up on.

nagilson commented on src/Tests/Microsoft.NET.Build.Containers.IntegrationTests/DockerRegistryManager.cs:
```diff
@@ -10,12 +10,15 @@ namespace Microsoft.NET.Build.Containers.IntegrationTests;
+    public const string Net8PreviewImageTag = "8.0-preview";
     public const string LocalRegistry = "localhost:5010";
-    public const string FullyQualifiedBaseImageDefault = $"{BaseImageSource}{BaseImage}:{Net6ImageTag}";
+    public const string FullyQualifiedBaseImageDefault = $"{BaseImageSource}{RuntimeBaseImage}:{Net8PreviewImageTag}";
+    public const string FullyQualifiedBaseImageAspNet = $"{BaseImageSource}{AspNetBaseImage}:{Net8PreviewImageTag}";
```
These are from https://github.com/dotnet/sdk/pull/30038/commits/72b31a43e8d1426380cac002fbff1178ff95d00a which was needed to get the tests to pass irrespective of this change I _think_, should we just put that in a separate PR @baronfel ?

nagilson commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildASelfContainedApp.cs:
```diff
@@ -383,6 +385,75 @@ public void It_can_publish_runtime_specific_apps_with_library_dependencies_self_
+                    var ns = project.Root.Name.Namespace;
+                    var propertyGroup = project.Root.Elements(ns + "PropertyGroup").First();
+                    propertyGroup.Add(new XElement(ns + "RuntimeIdentifier", runtimeIdentifier));
+                    propertyGroup.Add(new XElement(ns + "SelfContained", defineSelfContained ? "true" : ""));
```
How would we best do that? The project change needs to be within a lambda expression. I dont think there is a way to wrap an if statement inside of a lambda expression in particular.. and I am not aware of any C# language feature where we could pass a statement into the lambda thats a no-op lest we create a function for it such as `defineSelfContained ? propertyGroup.Add(...) : doNothing()`?

dsplaisted replied:
Unless I'm very confused you can have an if statement inside here no problem:

```c#
                    var ns = project.Root.Name.Namespace;
                    var propertyGroup = project.Root.Elements(ns + "PropertyGroup").First();
                    propertyGroup.Add(new XElement(ns + "RuntimeIdentifier", runtimeIdentifier));
                    if (defineSelfContained)
                    {
                        propertyGroup.Add(new XElement(ns + "SelfContained", "true"));
                    }
```

nagilson replied:
Nope, I was the very confused one. Lol

nagilson committed:
undo blazor wasm self contained test additions because they prepared for the break on pr 30700

nagilson committed:
Revert "undo blazor wasm self contained test additions because they prepared for the break on pr 30700"

This reverts commit 26550ed91f5eafb28e73d312737a739d7dd3b437.

nagilson committed:
respond to most pr feedback

nagilson committed:
WIP: Modify validate exe reference test to account for tfm

nagilson committed:
wip fix exe reference framework error but tfm isnt available in the  metadata coming from msbuild

nagilson committed:
Fix referenced exe error to be tfm gateD

nagilson commented on src/Tasks/Microsoft.NET.Build.Tasks/ValidateExecutableReferences.cs:
```diff
@@ -75,14 +75,6 @@ protected override void ExecuteCore()
-
-                    //  If the project is NOT RID agnostic, then a global RuntimeIdentifier will flow to it.
-                    //  If the project didn't explicitly specify a value for SelfContained, then this will
-                    //  set SelfContained to true
```
That is a really good point. I think I've fixed this and added tests for it.

nagilson committed:
Add selfcontained to blazor wasm library project because it still seems to need it

nagilson commented on src/Assets/TestProjects/BlazorHosted/blazorhosted/blazorhosted.csproj:
```diff
@@ -8,6 +8,11 @@
+  <!-- Based on .NET 8 breaking change: Add SelfContained if RuntimeIdentifier is defined to match old behavior.. -->
+  <PropertyGroup Condition="'$(RuntimeIdentifier)' != '' and '$(SelfContained)' == ''">
+    <SelfContained>true</SelfContained>
+  </PropertyGroup>
```
Unfortunately, it looks like there are still edge cases where Blazor would be broken so I am coordinating with them. I don't think we want to release this and break Blazor in the process so it will probably get delayed to preview 5 unless we can do it quickly.

cc @javiercn @danroth27

nagilson committed:
Add SelfContained to the TestProject Library and clean up the tests

nagilson committed:
Add SelfContained back into the blazor tests because they havent fully patched it it seems and so we have to delay this to preview 5 and id like to make the build green so we can merge it

nagilson committed:
fix remaining tests

nagilson committed:
update container tag to 8 becuase older tfms dont appear to work so I bumped the test tfm

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/ReferenceExeTests.cs:
```diff
@@ -240,8 +265,12 @@ public void ReferencedExeCanRunWhenReferencesExeWithSelfContainedMismatchForDiff
 
                 project.Root.Element(ns + "PropertyGroup")
                     .Add(XElement.Parse($@"<RuntimeIdentifier Condition=""'$(TargetFramework)' == '{ToolsetInfo.CurrentTargetFramework}'"">" + EnvironmentInfo.GetCompatibleRid() + "</RuntimeIdentifier>"));
+
+                project.Root.Element(ns + "PropertyGroup")
+                    .Add(new XElement(ns + "SelfContained", "true"));
```
To test the intended scenario, `SelfContained` should only be set to true for the non `windows` TargetFramework.  Here I've moved both properties into a new PropertyGroup with the appropriate condition.

```suggestion
                var propertyGroup = new XElement(ns + "PropertyGroup",
                    new XAttribute("Condition", $"'$(TargetFramework)' == '{ToolsetInfo.CurrentTargetFramework}'"));

                propertyGroup.Add(new XElement(ns + "RuntimeIdentifier", EnvironmentInfo.GetCompatibleRid()));
                propertyGroup.Add(new XElement(ns + "SelfContained", "true"));

                project.Root.Add(propertyGroup);
```

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ValidateExecutableReferences.cs:
```diff
@@ -54,6 +57,11 @@ protected override void ExecuteCore()
+                // We can only access TargetFrameworks and NearestTargetFramework to find the referenced project "TargetFramework".
+                // We rely on the nearest one because it will pick the lowest 'most-compatible' tfm for the two projects.
+                // Since 'younger' TFMs are the ones that would error and are generally also what gets picked as 'most-copmaptible' we can use it.
+                var referencedProjectTargetFramework = NuGetFramework.Parse(nearestTargetFramework);
```
We should not be parsing the TargetFramework here, that will not work well with aliased TargetFrameworks (see [here](https://github.com/NuGet/Home/issues/5154) and [here](https://github.com/NuGet/Home/pull/12124).

I think what we should instead be doing is getting the index of the NearestTargetFramework from the TargetFrameworks metadata, and then retrieving the value from the same index in the TargetFrameworkMonikers metadata.  You can then parse that moniker with the [FrameworkName](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.versioning.frameworkname.-ctor?view=net-8.0#system-runtime-versioning-frameworkname-ctor(system-string)) class, and check that the `Identifier` is `.NETCoreApp` and the `Version` is less than 8.

Here's a sample of what the data for a multi-targeted referenced project looks like from a binlog:

```
../ReferencedProject/ReferencedProject.csproj
    ReferenceSourceTarget = ProjectReference
    MSBuildSourceProjectFile = c:\git\dotnet-sdk-main\artifacts\tmp\Debug\ReferencedExe---C4AC8564\ReferencedProject\ReferencedProject.csproj
    HasSingleTargetFramework = false
    SetTargetFramework = TargetFramework=net8.0
    ReferenceOutputAssembly = true
    IsVcxOrNativeProj = false
    OutputItemType = 
    TargetPlatformMonikers = None;Windows,Version=7.0
    BuildReference = true
    Platform = AnyCPU
    AdditionalPropertiesFromProject = <AdditionalProjectProperties>
  <TargetFramework Name="net8.0">
    <SelfContained>true</SelfContained>
    <_IsExecutable>true</_IsExecutable>
    <IsRidAgnostic>false</IsRidAgnostic>
    <ShouldBeValidatedAsExecutableReference></ShouldBeValidatedAsExecutableReference>
    <_SelfContainedWasSpecified>true</_SelfContainedWasSpecified>
  </TargetFramework>
  <TargetFramework Name="net8.0-windows">
    <SelfContained>false</SelfContained>
    <_IsExecutable>true</_IsExecutable>
    <IsRidAgnostic>false</IsRidAgnostic>
    <ShouldBeValidatedAsExecutableReference></ShouldBeValidatedAsExecutableReference>
    <_SelfContainedWasSpecified></_SelfContainedWasSpecified>
  </TargetFramework>
</AdditionalProjectProperties>
    TargetFrameworkMonikers = .NETCoreApp,Version=v8.0;.NETCoreApp,Version=v8.0
    TargetFrameworks = net8.0;net8.0-windows
    IsRidAgnostic = false
    MSBuildSourceTargetName = GetTargetFrameworks
    NearestTargetFramework = net8.0
    Platforms = AnyCPU
    Targets = 
    OriginalItemSpec = ../ReferencedProject/ReferencedProject.csproj
```

nagilson replied:
Thanks, I didn't consider this and pushed the change. I must admit I don't really get the difference between this and the TargetFrameworkElement retrived using `UseAttributeForTargetFrameworkInfoPropertyNames` and I didn't leverage that property. Please let me know if we need to use that as I don't really see what it is being used for

nagilson committed:
remove the binlog i forgot in the test

Co-authored-by: Daniel Plaisted <dsplaisted@gmail.com>

nagilson committed:
Change Test for ReferenceExe to not Specify SelfContained if RuntimeIdentifier is Undesired

We do this as SelfContained can imply RuntimeIdentifier and before RID used to imply SelfContained, and we want this beahvior to be coupled in this  test

Co-authored-by: Daniel Plaisted <dsplaisted@gmail.com>

nagilson committed:
Account for targetframework aliases while parsing the target framework to error in older selfcontained exe reference scenarios

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ValidateExecutableReferences.cs:
```diff
@@ -76,10 +94,14 @@ protected override void ExecuteCore()
                     //  If the project didn't explicitly specify a value for SelfContained, then this will
                     //  set SelfContained to true
-                    if (runtimeIdentifierIsGlobalProperty && !referencedProjectHadSelfContainedSpecified)
+                    if (RuntimeIdentifierInfersSelfContained && runtimeIdentifierIsGlobalProperty && !referencedProjectHadSelfContainedSpecified)
```
I think this needs to refer to the local variable, rather than the property which doesn't ever seem to be set.

```suggestion
                    if (runtimeIdentifierInfersSelfContained && runtimeIdentifierIsGlobalProperty && !referencedProjectHadSelfContainedSpecified)
```

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ValidateExecutableReferences.cs:
```diff
@@ -76,10 +94,14 @@ protected override void ExecuteCore()
                     }
 
+                    //  We need to check if referenced project will become SelfContained because of its RuntimeIdentifier. This only happens on TargetFrameworks less than net8.0.
+                    var sdkVersionWhereRuntimeIdentifierNoLongerInfersSelfContained = new FrameworkName(".NETCoreApp,Version=v8.0");
+                    bool runtimeIdentifierInfersSelfContained = referencedProjectTargetFramework != null && referencedProjectTargetFramework.Identifier == ".NETCoreApp" && referencedProjectTargetFramework.Version < sdkVersionWhereRuntimeIdentifierNoLongerInfersSelfContained.Version;
```
I think this can be simplified:

```suggestion
                    bool runtimeIdentifierInfersSelfContained = referencedProjectTargetFramework != null && referencedProjectTargetFramework.Identifier == ".NETCoreApp" && referencedProjectTargetFramework.Version.Major < 8;
```

nagilson committed:
Refer to the new variable we created instead of the old one that shouldve been deleted

Co-authored-by: Daniel Plaisted <dsplaisted@gmail.com>

nagilson committed:
Use an integer to compare versions instead of comparing with a created version class to simplify code

Co-authored-by: Daniel Plaisted <dsplaisted@gmail.com>

nagilson committed:
remove unneeeded boolean as we took a different approach to consider tfm aliases

nagilson committed:
Merge remote-tracking branch 'upstream/main' into nagilson-rid-no-more-sc-2.0

nagilson committed:
fix merge conflict that vscode did not save for some reason

nagilson commented:
Here's the breaking change doc: https://github.com/dotnet/docs/issues/33726
